osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs

```diff
@@ -39,6 +39,8 @@ public class ComposeBlueprintContainer : EditorBlueprintContainer
         private PlacementBlueprint currentPlacement;
         private InputManager inputManager;
 
+        private DragEvent lastDragEvent;
```

Comment:
How about moving this to the base class to avoid calling `OnDrag` from `Update`? I feel a touch uneasy calling a framework input flow manually like this.

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
index b15cecd506..d7cabc6399 100644
--- a/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs
@@ -170,6 +170,8 @@ protected override void OnMouseUp(MouseUpEvent e)
             finishSelectionMovement();
         }
 
+        private MouseButtonEvent lastDragEvent;
+
         protected override bool OnDragStart(DragStartEvent e)
         {
             if (e.Button == MouseButton.Right)
@@ -182,6 +184,7 @@ protected override bool OnDragStart(DragStartEvent e)
                 return true;
             }
 
+            lastDragEvent = e;
             DragBox.HandleDrag(e);
             DragBox.Show();
             return true;
@@ -189,15 +192,7 @@ protected override bool OnDragStart(DragStartEvent e)
 
         protected override void OnDrag(DragEvent e)
         {
-            if (e.Button == MouseButton.Right)
-                return;
-
-            if (DragBox.State == Visibility.Visible)
-            {
-                DragBox.HandleDrag(e);
-                UpdateSelectionFromDragBox();
-            }
-
+            lastDragEvent = e;
             moveCurrentSelection(e);
         }
 
@@ -215,6 +210,17 @@ protected override void OnDragEnd(DragEndEvent e)
             DragBox.Hide();
         }
 
+        protected override void Update()
+        {
+            base.Update();
+
+            if (IsDragged && DragBox.State == Visibility.Visible)
+            {
+                DragBox.HandleDrag(lastDragEvent);
+                UpdateSelectionFromDragBox();
+            }
+        }
+
         /// <summary>
         /// Called whenever a drag operation completes, before any change transaction is committed.
         /// </summary>
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs
index 31990bfd35..f0a9d70df3 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs
@@ -29,7 +29,6 @@ internal class TimelineBlueprintContainer : EditorBlueprintContainer
         [Resolved(CanBeNull = true)]
         private Timeline timeline { get; set; }
 
-        private DragEvent lastDragEvent;
         private Bindable<HitObject> placement;
         private SelectionBlueprint<HitObject> placementBlueprint;
 
@@ -105,18 +104,8 @@ protected override void OnDrag(DragEvent e)
             base.OnDrag(e);
         }
 
-        protected override void OnDragEnd(DragEndEvent e)
-        {
-            base.OnDragEnd(e);
-            lastDragEvent = null;
-        }
-
         protected override void Update()
         {
-            // trigger every frame so drags continue to update selection while playback is scrolling the timeline.
-            if (lastDragEvent != null)
-                OnDrag(lastDragEvent);
-
             if (Composer != null && timeline != null)
             {
                 Composer.Playfield.PastLifetimeExtension = timeline.VisibleRange / 2;
@@ -199,11 +188,6 @@ bool shouldBeSelected(HitObject hitObject)
 
         private void handleScrollViaDrag(DragEvent e)
         {
-            lastDragEvent = e;
-
-            if (lastDragEvent == null)
-                return;
-
             if (timeline != null)
             {
                 var timelineQuad = timeline.ScreenSpaceDrawQuad;

```

Reply:
Done, with some more code change to support timeline blueprint container scrolling while dragging hit objects.